src/PureScript/CST/ModuleGraph.purs

```diff
@@ -0,0 +1,120 @@
+module PureScript.CST.ModuleGraph
+  ( moduleGraph
+  , sortModules
+  ) where
+
+import Prelude
+
+import Data.Array as Array
+import Data.Bifunctor (bimap)
+import Data.Either (Either(..))
+import Data.Foldable (all, foldl)
+import Data.List (List(..))
+import Data.List as List
+import Data.Map (Map)
+import Data.Map as Map
+import Data.Maybe (Maybe(..), fromMaybe, maybe)
+import Data.Set (Set)
+import Data.Set as Set
+import Data.Tuple (Tuple(..))
+import Control.Alt (alt)
+import PureScript.CST.Types (ImportDecl(..), ModuleHeader(..), ModuleName, Name(..))
+
+type Graph a = Map a (Set a)
+
+moduleGraph :: forall e. Array (ModuleHeader e) -> Graph ModuleName
+moduleGraph = Map.fromFoldable <<< map go
+  where
+  go (ModuleHeader { name: Name { name }, imports }) =
+    Tuple name (Set.fromFoldable (map getImportName imports))
+
+  getImportName (ImportDecl { "module": Name { name } }) = name
+
+sortModules :: forall e. Array (ModuleHeader e) -> Either (Array (ModuleHeader e)) (Array (ModuleHeader e))
```

Comment:
What do you think about making this `Either` something like:

```purescript
data ModuleSort e
  = Sorted (Array (ModuleHeader e))
  | CycleDetected (Array (ModuleHeader e))
```

Reply:
Yeah, I think that makes it easier to consume - I'll make the change.